util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+		l("timed out")
+		return 1, e
+	}
+	if waitCode != windows.WAIT_OBJECT_0 {
+		// This should not be possible; log
+		return 1, fmt.Errorf("unexpected state from WaitForSingleObject: %d", waitCode)
+	}
+
+	l("Process completed, obtaining exit code")
+	var exitCode uint32
+	if err := windows.GetExitCodeProcess(pi.Process, &exitCode); err != nil {
+		return 1, err
+	}
+	l("Returning exit code %d", exitCode)
+	return exitCode, nil
+}
+
+func startProcessInSessionInternal(logf logger.Logf, sessID SessionID, cmdLineInfo CommandLineInfo, extraFlags uint32) (*windows.ProcessInformation, error) {
+	logf("Validating cmdLineInfo")
+	if err := cmdLineInfo.Validate(); err != nil {
+		return nil, err
+	}
+
+	logf("Obtaining and validating token for TS session %d", sessID.TSSession)
+	token, err := sessID.OpenToken()
+	if err != nil {
+		return nil, fmt.Errorf("(*SessionID).OpenToken: %w", err)
+	}
+	defer token.Close()
+
+	logf("Resolving exe and command line paths")
+	exePath16, commandLine16, commandLineStr, err := cmdLineInfo.Resolve()
+	if err != nil {
+		return nil, fmt.Errorf("(*CommandLineInfo).Resolve(): %w", err)
+	}
+
+	logf("Resolving working directory")
+	wd16, err := windows.UTF16PtrFromString(filepath.Dir(cmdLineInfo.ExePath))
+	if err != nil {
+		return nil, fmt.Errorf("UTF16PtrFromString(wd): %w", err)
+	}
+
+	logf("Building environment block")
+	env, err := token.Environ(false)
+	if err != nil {
+		return nil, fmt.Errorf("token environment: %w", err)
+	}
+	env16 := newEnvBlock(env)
+
+	logf("Enabling privileges")
+	// The privileges in privNames are required for CreateProcessAsUser to be
+	// able to start processes as other users in other logon sessions.
+	privNames := []string{
+		"SeAssignPrimaryTokenPrivilege",
+		"SeIncreaseQuotaPrivilege",
+	}
+	dropPrivs, err := EnableCurrentThreadPrivileges(privNames)
+	if err != nil {
+		return nil, fmt.Errorf("EnableCurrentThreadPrivileges(%#v): %w", privNames, err)
+	}
+	defer dropPrivs()
+
+	logf("Calling CreateProcessAsUser: %s", commandLineStr)
+	createFlags := extraFlags | windows.CREATE_UNICODE_ENVIRONMENT | windows.DETACHED_PROCESS
+	si := windows.StartupInfo{
+		Cb:      uint32(unsafe.Sizeof(windows.StartupInfo{})),
+		Desktop: defaultDesktop,
+	}
+	var pi windows.ProcessInformation
+	if err := windows.CreateProcessAsUser(token, exePath16, commandLine16, nil, nil,
+		false, createFlags, env16, wd16, &si, &pi); err != nil {
+		return nil, fmt.Errorf("CreateProcessAsUser: %w", err)
+	}
+	logf("Succeeded")
+	return &pi, nil
+}
+
+func newEnvBlock(env []string) *uint16 {
+	// Intentionally using bytes.Buffer here because we're writing nul bytes (the standard library does this too).
+	var buf bytes.Buffer
+	for _, v := range env {
+		buf.WriteString(v)
+		buf.WriteByte(0)
+	}
+	if buf.Len() == 0 {
+		// So that we end with a double-null in the empty env case
+		buf.WriteByte(0)
+	}
+	buf.WriteByte(0)
+	return unsafe.SliceData(utf16.Encode([]rune(string(buf.Bytes()))))
```

Comment:
Do utf-8 strings allow null characters? Is converting through string technically against the rules?

Reply:
okay, seems reasonable